[HUDI-7144] Build storage partition stats index and use it for data skipping

codope opened the pull request:
### Change Logs

Build storage partition stats index and use it for data skipping. Main changes are as follows:

- Index is saved as another partition in the metadata table.
- Each index entry is a key-value, where key is the hash(columnName).concat(hash(partitionName)), and value is the stats.
- New configs in `HoodieMetadataConfig` and the writer changes are in `HoodieBackedTableMetadataWriter` with some util methods in `HoodieTableMetadataUtil`
- On the read path, main changes are in `HoodieFileIndex`. First, the partition pruning happens as usual, then depending on data filters, data can be skipped further if partition stats index is available.

### Impact

Stats aggregated by storage partition. Efficient data skipping. Meta sync need not sync the partition metadata. Queries will use the index while planning in the driver.


### Risk level (write none, low medium or high below)

medium

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1901,4 +1909,162 @@ private static Path filePath(String basePath, String partition, String filename)
+    if (columnsToIndex.isEmpty()) {
+      return engineContext.emptyHoodieData();
+    }
+    LOG.info(String.format("Indexing %d columns for partition stats index", columnsToIndex.size()));
```
info?

codope replied:
changed to debug mode.. was more for lcoal testing.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/metadata/MetadataPartitionType.java:
```diff
@@ -26,6 +26,7 @@
  */
 public enum MetadataPartitionType {
   FILES(HoodieTableMetadataUtil.PARTITION_NAME_FILES, "files-"),
+  PARTITION_STATS(HoodieTableMetadataUtil.PARTITION_NAME_PARTITION_STATS, "part-stats-"),
```
rename: partition-stats

codope replied:
done

vinothchandar commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/PartitionStatsIndexSupport.scala:
```diff
@@ -0,0 +1,71 @@
+    val metadataRecords: HoodieData[HoodieRecord[HoodieMetadataPayload]] =
+      metadataTable.getRecordsByKeyPrefixes(encodedTargetColumnNames.asJava, HoodieTableMetadataUtil.PARTITION_NAME_PARTITION_STATS, shouldReadInMemory)
+    val columnStatsRecords: HoodieData[HoodieMetadataColumnStats] =
+      // NOTE: Explicit conversion is required for Scala 2.11
```
side note: I think we can drop support for 2.11?

codope replied:
yeah there are bunch of other places we do things just to be compatible with 2.11. I'll track it separately.

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/PartitionStatsIndexTestBase.scala:
```diff
@@ -0,0 +1,259 @@
+    metaClient
+  }
+
+  protected def rollbackLastInstant(hudiOpts: Map[String, String]): HoodieInstant = {
```
is nt there some test util class we can reuse for rolling back the last instant? general comment is on - can we cut down the amount of additional test code and reuse? or we need specific things here?

codope replied:
Did not find one for this method. But, i get your point. I also see some near-duplicate methods in subclasses of `HoodieSparkClientTestBase`. I will refactor those separately.

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/PartitionStatsIndexTestBase.scala:
```diff
@@ -0,0 +1,259 @@
+    "hoodie.upsert.shuffle.parallelism" -> "4",
+    HoodieWriteConfig.TBL_NAME.key -> "hoodie_test",
+    RECORDKEY_FIELD.key -> "_row_key",
+    PARTITIONPATH_FIELD.key -> "partition",
```
lets change this test to have multiple partition fields, so we have a multi-level /x/y/z like partition structure

codope replied:
done - `PARTITIONPATH_FIELD.key -> "partition,trip_type"`

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndex.scala:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  @Test
+  def testPartitionStatsIndexWithSQL(): Unit = {
```
lets cover CTAS, followed by insert, update, merge, delete statements.

codope replied:
added a test

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/PartitionStatsIndexTestBase.scala:
```diff
@@ -0,0 +1,259 @@
+    val partitionStats = partitionStatsIndex.loadColumnStatsIndexRecords(List("partition"), shouldReadInMemory = true).collectAsList()
+    assertEquals(0, partitionStats.size())
+    val prevDf = mergedDfList.last.drop("tip_history")
+    val nonMatchingRecords = readDf.drop("_hoodie_commit_time", "_hoodie_commit_seqno", "_hoodie_record_key",
```
is nt there a method for this to drop the meta fields from a DF?

codope replied:
yeah there is.. reusing that now.

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/PartitionStatsIndexTestBase.scala:
```diff
@@ -0,0 +1,259 @@
+    }
+    val latestBatchDf = spark.read.json(spark.sparkContext.parallelize(latestBatch, 2))
+    latestBatchDf.cache()
+    latestBatchDf.write.format("org.apache.hudi")
```
does `.partitionBy(..)` work? Same question for a CTAS statement with partition columns.

vinothchandar reviewed (changes requested):
Left some comments on cases that I think we should cover out of the gate and have it all passing tests. 

Summarizing by understanding here, to get on the same page. When the partition stats index is turned on, we essentially track merged statistics across all files within a storage partition, for each of the partition columns. on the query side, we try to see if any of the storage partitions can be simply pruned out based on columns in the query and the new stats. 

So, there should be no semantic differences between this and what hive-style partitioning will achieve. i.e queries written on one of the partition columns and a combination of them etc. Love to see complex test cases like those and confirm this is indeed good

I can do line-by-line detailed review once these are ressolved.

vinothchandar commented:
I made one skim of the changes. Can you reply on my previous review comments? esp on tests.

codope commented on hudi-common/src/main/java/org/apache/hudi/common/config/HoodieMetadataConfig.java:
```diff
@@ -330,6 +330,27 @@ public final class HoodieMetadataConfig extends HoodieConfig {
       .sinceVersion("1.0.0")
       .withDocumentation("Parallelism to use, when generating functional index.");
 
+  public static final ConfigProperty<Boolean> ENABLE_METADATA_INDEX_PARTITION_STATS = ConfigProperty
```
done

codope commented on hudi-common/src/main/java/org/apache/hudi/common/util/ParquetUtils.java:
```diff
@@ -454,6 +454,51 @@ private static BigDecimal extractDecimal(Object val, DecimalMetadata decimalMeta
+   *
+   * @param fileRanges List of column range statistics for each file in a partition
+   */
+  public <T extends Comparable<T>> HoodieColumnRangeMetadata<T> getColumnRangeInPartition(@Nonnull List<HoodieColumnRangeMetadata<T>> fileRanges) {
```
Moved to BaseFileUtils and added unit test.

codope commented:
> I made one skim of the changes. Can you reply on my previous review comments? esp on tests.

@vinothchandar Thanks for the review. I have addressed all your comments. The test was passing two commits ago. I am looking into the failures. But, it's good to review.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1901,4 +1910,162 @@ private static Path filePath(String basePath, String partition, String filename)
+                                                                                       HoodieTableMetaClient datasetMetaClient,
+                                                                                       List<String> columnsToIndex,
+                                                                                       boolean isDeleted) {
+    String filePartitionPath = filePath.startsWith("/") ? filePath.substring(1) : filePath;
```
what are the cases where we would and would not get a "/". or is this just defensive ?

codope replied:
This `filePartitionPath` is a combination of partition name + "/" + file name. For non-partitioned tables, we can have "/" in the beginning.

vinothchandar reviewed (commented):
Left some comments. Will need to step through to understand how partition columns are mapped automatically on the writer side into target column stats index cols.

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndex.scala:
```diff
@@ -0,0 +1,188 @@
+      .mode(SaveMode.Overwrite)
+      .save(basePath)
+
+    val snapshot0 = spark.read.format("org.apache.hudi").options(hudiOpts).load(basePath)
```
this query is not trying to prune partitions?

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndexWithSql.scala:
```diff
@@ -0,0 +1,106 @@
+      assertResult(WriteOperationType.BULK_INSERT) {
+        getLastCommitMetadata(spark, tablePath).getOperationType
+      }
+      checkAnswer(s"select id, name, price, ts, cast(dt as string) from $tableName")(
```
this or any other query below, does not seem to be testing partition pruning? Verifying partition field value is great, but We should have tests that does that, right?

codope replied:
I will add a test to assert partition pruning based on plan.

yihua replied:
@codope is this done?  I don't seem to see the validation on the partition prunning in the tests in this class.

vinothchandar reviewed (changes requested):
Please help me understand if I am missing sth around tests

vinothchandar commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala:
```diff
@@ -346,6 +352,12 @@ case class HoodieFileIndex(spark: SparkSession,
       Option.apply(recordLevelIndex.getCandidateFiles(getAllFiles(), recordKeys))
+    } else if (recordKeys.nonEmpty && partitionStatsIndex.isIndexAvailable && !queryFilters.isEmpty) {
+      val prunedFileNames = getPrunedFileNames(prunedPartitionsAndFileSlices)
+      val shouldReadInMemory = partitionStatsIndex.shouldReadInMemory(this, queryReferencedColumns)
```
Can you please point me to the piece of code/line numbers where we identify a referenced column as a partitioned column? (on read side) 
and how `.partitionBy("a,b,c")` turns into partition stats on the metadata write side?

That part is a still unclear to me.

codope replied:
First thing I want to clarify is that partition stats are collected only when column stats is enabled, and for only those columns for which column stats is enabled.

Write path: When user will do `.partitionBy("a,b,c")`, then the logic is similar to column stats. We use the commit metadata  and convert that to partition stats. This happens in `HoodieTableMetadataUtil.convertMetadataToPartitionStatsRecords`. The difference from column stats is that the stats are aggregated by partition value in `BaseFileUtils.getColumnRangeInPartition`.

Read path: `queryReferencedColumns` here contain data filters. Partition pruning based on partition filters has already happened one level above.

vinothchandar reviewed (changes requested):
Bunch more comments.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1915,4 +1924,162 @@ private static Path filePath(String basePath, String partition, String filename)
+    return engineContext.parallelize(partitionInfoList, parallelism).flatMap(partitionFiles -> {
+      final String partitionName = partitionFiles.getRelativePath();
+      Stream<HoodieColumnRangeMetadata<Comparable>> partitionStatsRangeMetadata = partitionFiles.getFileNameToSizeMap().keySet().stream()
+          .map(fileName -> getFileStatsRangeMetadata(partitionName, partitionName + "/" + fileName, dataTableMetaClient, columnsToIndex, false))
```
are we reading all the column stats again to generate partition stats? I think we should avoid doing this extra work and piggy back off the column stats index values?

codope replied:
This is called one time at the time of intialization of partiiton stats. I thought if users enabled partition stats only (no column stats) then we need to make this call. Nevertheless, we can avoid extra work in most cases with minimal branching.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1915,4 +1924,162 @@ private static Path filePath(String basePath, String partition, String filename)
+    LOG.debug(String.format("Indexing %d columns for partition stats index", columnsToIndex.size()));
+    // Create records for MDT
+    int parallelism = Math.max(Math.min(partitionInfoList.size(), recordsGenerationParams.getPartitionStatsIndexParallelism()), 1);
+    return engineContext.parallelize(partitionInfoList, parallelism).flatMap(partitionFiles -> {
```
rename: partitionFiles -> partitionInfo

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1915,4 +1924,162 @@ private static Path filePath(String basePath, String partition, String filename)
+      final String partitionName = partitionFiles.getRelativePath();
+      Stream<HoodieColumnRangeMetadata<Comparable>> partitionStatsRangeMetadata = partitionFiles.getFileNameToSizeMap().keySet().stream()
+          .map(fileName -> getFileStatsRangeMetadata(partitionName, partitionName + "/" + fileName, dataTableMetaClient, columnsToIndex, false))
+          .map(BaseFileUtils::getColumnRangeInPartition);
```
IIUC are n't you passing all column ranges for a given file into this? Would n't that merge ranges across different columns within a given file?

codope replied:
This is fixed now and added a test `testConvertFilesToPartitionStatsRecords`

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -67,6 +70,50 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+   * @param fileRanges List of column range statistics for each file in a partition
+   */
+  public static <T extends Comparable<T>> HoodieColumnRangeMetadata<T> getColumnRangeInPartition(@Nonnull List<HoodieColumnRangeMetadata<T>> fileRanges) {
+    if (fileRanges.size() == 1) {
```
I think this is column range, and not multiple files? i.e list has an entry for each column for a single file. no?

codope replied:
Yes, fixed the logic and added ut in `testGetColumnRangeInPartition`

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -67,6 +70,50 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+    // There are multiple files. Compute min(file_mins) and max(file_maxs)
+    return fileRanges.stream()
+        .sequential()
+        .reduce(BaseFileUtils::mergeRanges).get();
```
we should be merging across files on a given column

codope replied:
yea, apologies, current logic is incorrect. i'm going to fix it.

codope replied:
fixed

vinothchandar reviewed (changes requested):
Please clarify the latest comments.

codope commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -64,6 +67,51 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+    }
+
+    return HoodieColumnRangeMetadata.create(
+        one.getFilePath(),
```
done

codope commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java:
```diff
@@ -641,6 +642,36 @@ public static Stream<HoodieRecord> createColumnStatsRecords(String partitionName
     });
   }
 
+  public static Stream<HoodieRecord> createPartitionStatsRecords(String partitionName,
```
Changed to partitionPath

codope commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1901,4 +1910,162 @@ private static Path filePath(String basePath, String partition, String filename)
+
+  public static HoodieData<HoodieRecord> convertFilesToPartitionStatsRecords(HoodieEngineContext engineContext,
+                                                                             List<DirectoryInfo> partitionInfoList,
+                                                                             MetadataRecordsGenerationParams recordsGenerationParams) {
```
Removed usage of `MetadataRecordsGenerationParams`. Eventually, we should just get rid of the pojo.

codope commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1901,4 +1910,162 @@ private static Path filePath(String basePath, String partition, String filename)
     }
   }
+
+  public static HoodieData<HoodieRecord> convertFilesToPartitionStatsRecords(HoodieEngineContext engineContext,
```
no, we need to specify the partition field in target columns config.

codope replied:
We can avoid this extra step though.

codope commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala:
```diff
@@ -346,6 +352,12 @@ case class HoodieFileIndex(spark: SparkSession,
       Option.empty
     } else if (recordKeys.nonEmpty) {
       Option.apply(recordLevelIndex.getCandidateFiles(getAllFiles(), recordKeys))
+    } else if (recordKeys.nonEmpty && partitionStatsIndex.isIndexAvailable && !queryFilters.isEmpty) {
```
not necessary.. removed.

codope commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/PartitionStatsIndexSupport.scala:
```diff
@@ -0,0 +1,71 @@
+  extends ColumnStatsIndexSupport(spark, tableSchema, metadataConfig, metaClient, allowCaching) {
+
+  @transient private lazy val engineCtx = new HoodieSparkEngineContext(new JavaSparkContext(spark.sparkContext))
+  @transient private lazy val metadataTable: HoodieTableMetadata =
```
Refactored as per suggestion

codope commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/PartitionStatsIndexSupport.scala:
```diff
@@ -0,0 +1,71 @@
+    HoodieTableMetadata.create(engineCtx, metadataConfig, metaClient.getBasePathV2.toString)
+
+  override def isIndexAvailable: Boolean = {
+    checkState(metadataConfig.enabled, "Metadata Table support has to be enabled")
```
not necessary as we are checking whether table config contains partition stats in the list of of available indexes.

codope commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndex.scala:
```diff
@@ -0,0 +1,188 @@
+    val records = recordsToStrings(dataGen.generateInserts("000", 100)).toList
+    val inputDF = spark.read.json(spark.sparkContext.parallelize(records, 2))
+
+    inputDF.write.partitionBy("partition").format("hudi")
```
this is resolved now. Had to url encode the slashes.
<img width="890" alt="Screenshot 2024-04-10 at 10 19 15 PM" src="https://github.com/apache/hudi/assets/16440354/db7dbe15-a47a-4f1c-8035-f6c2c75d20d1">

codope commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -64,6 +67,51 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+    final T minValue;
+    final T maxValue;
+    if (one.getMinValue() != null && another.getMinValue() != null) {
+      minValue = one.getMinValue().toString().compareTo(another.getMinValue().toString()) < 0 ? one.getMinValue() : another.getMinValue();
```
No, lexicographical ordering isn't correct for all types. Since the type already extends Comparable, i've removed `toString`.

yihua commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/table/upgrade/TestUpgradeDowngrade.java:
```diff
@@ -557,7 +558,8 @@ public void testDowngradeSixToFiveShouldDeleteRecordIndexPartition() throws Exce
         PARTITION_NAME_BLOOM_FILTERS,
         PARTITION_NAME_RECORD_INDEX,
-        PARTITION_NAME_FUNCTIONAL_INDEX_PREFIX
+        PARTITION_NAME_FUNCTIONAL_INDEX_PREFIX,
```
Should this list be generated from the production code, i.e., list of supported MDT partitions?  Also, do we need to upgrade the table version?  I think master branch is still on table version 6, the same as 0.14.0 release.

codope replied:
I think we will eventually need to upgrade the table version for 1.0. But, this PR does not do any incompatible schema changes, so we should be good for now. HUDI-7665 to track.

yihua commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1580,69 +1595,4 @@ public boolean isInitialized() {
-   * A {@code PartitionFileInfo} object saves the name of the partition and various properties requires of each file
-   * required for initializing the metadata table. Saving limited properties reduces the total memory footprint when
-   * a very large number of files are present in the dataset being initialized.
-   */
```
So this is no longer needed?

codope replied:
Needed, but extracted for reuse.

yihua commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -67,6 +70,50 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+        .reduce(BaseFileUtils::mergeRanges).get();
+  }
+
+  private static  <T extends Comparable<T>> HoodieColumnRangeMetadata<T> mergeRanges(HoodieColumnRangeMetadata<T> one,
```
```suggestion
  private static <T extends Comparable<T>> HoodieColumnRangeMetadata<T> mergeRanges(HoodieColumnRangeMetadata<T> one,
```

yihua commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -67,6 +70,50 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+        .reduce(BaseFileUtils::mergeRanges).get();
+  }
+
+  private static  <T extends Comparable<T>> HoodieColumnRangeMetadata<T> mergeRanges(HoodieColumnRangeMetadata<T> one,
```
Do we already have such method for merging column ranges for col stats and could we reuse that?

codope replied:
Yes in `ParquetUtils`, some portion of this logic is duplicate. I have extracted to a common method in `BaseFileUtils` now.

yihua commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java:
```diff
@@ -104,13 +104,14 @@ public class HoodieMetadataPayload implements HoodieRecordPayload<HoodieMetadata
+  private static final int METADATA_TYPE_COLUMN_STATS = 3;
+  private static final int METADATA_TYPE_BLOOM_FILTER = 4;
   private static final int METADATA_TYPE_RECORD_INDEX = 5;
+  private static final int METADATA_TYPE_PARTITION_STATS = 6;
```
Maybe we should add enum and guarantee the ordering, and automatically assign the type ID?

codope replied:
Hmm.. yes we could. Just to be clear, you're suggesting no schema change right? As in we will still keep the int value in the records, and not enum.

yihua replied:
Right, no schema change; just code restructuring maintaining the same ID, i.e., add `getId` in `MetadataPartitionType`.  You can refactor this part in a separate PR.

yihua commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java:
```diff
@@ -651,6 +652,36 @@ public static Stream<HoodieRecord> createColumnStatsRecords(String partitionName
+    });
+  }
+
+  public static String getPartitionStatsIndexKey(String partitionPath, HoodieColumnRangeMetadata<Comparable> columnRangeMetadata) {
```
nit: have the column name as the method argument directly to avoid confusion?

yihua commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala:
```diff
@@ -351,6 +356,12 @@ case class HoodieFileIndex(spark: SparkSession,
+    } else if (partitionStatsIndex.isIndexAvailable && queryFilters.nonEmpty) {
+      val prunedFileNames = getPrunedFileNames(prunedPartitionsAndFileSlices)
+      val shouldReadInMemory = partitionStatsIndex.shouldReadInMemory(this, queryReferencedColumns)
+      partitionStatsIndex.loadTransposed(queryReferencedColumns, shouldReadInMemory) { transposedColStatsDF =>
```
Could we avoid transposing here as it has overhead and directly leverage the index entry for filtering?

codope replied:
This logic is now refactored and updated.

codope commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -224,6 +224,10 @@ private void enablePartitions() {
     }
+    if ((metadataConfig.isPartitionStatsIndexEnabled() && !metadataConfig.getColumnsEnabledForColumnStatsIndex().isEmpty())
+        || dataMetaClient.getTableConfig().isMetadataPartitionAvailable(PARTITION_STATS)) {
+      this.enabledPartitionTypes.add(PARTITION_STATS);
```
Done as part of https://github.com/apache/hudi/commit/9d689fd68f544cbd2424790770cc2673701be928

vinothchandar reviewed (commented):
Submitting a pending review, that I. had hanging around.

codope commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -67,6 +70,50 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+   */
+  public static <T extends Comparable<T>> HoodieColumnRangeMetadata<T> getColumnRangeInPartition(@Nonnull List<HoodieColumnRangeMetadata<T>> fileRanges) {
+    if (fileRanges.size() == 1) {
+      // Only one parquet file, we can just return that range.
```
done. added more UTs in `TestBaseFileUtils`

codope commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java:
```diff
@@ -651,6 +652,36 @@ public static Stream<HoodieRecord> createColumnStatsRecords(String partitionName
+    });
+  }
+
+  public static String getPartitionStatsIndexKey(String partitionPath, HoodieColumnRangeMetadata<Comparable> columnRangeMetadata) {
```
this method is covered in new tests added in `TestHoodieMetadataPayload` - we test merging two partition stats payloads, with and without delete.

codope commented on hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/functional/TestDataSkippingWithMORColstats.java:
```diff
@@ -318,6 +318,7 @@ private Map<String, String> getOptions() {
     options.put(HoodieMetadataConfig.ENABLE.key(), "true");
     options.put(HoodieMetadataConfig.ENABLE_METADATA_INDEX_COLUMN_STATS.key(), "true");
     options.put(HoodieMetadataConfig.COLUMN_STATS_INDEX_FOR_COLUMNS.key(), "trip_type");
+    options.put(HoodieMetadataConfig.ENABLE_METADATA_INDEX_PARTITION_STATS.key(), "false");
```
this class has some unit tests that deliberately corrupt parquet file by replacing a valid file with file of 0 bytes. In that case, keeping partition stats enabled throws error. To avoid that, I had disabled. In the latest commit, I have disabled only for tests that simulate corrupt parquet file as mentioned above, and not the whole test suite.

yihua commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -372,6 +379,10 @@ public static Map<MetadataPartitionType, HoodieData<HoodieRecord>> convertMetada
               dataMetaClient, isColumnStatsIndexEnabled, columnStatsIndexParallelism, targetColumnsForColumnStatsIndex);
       partitionToRecordsMap.put(MetadataPartitionType.COLUMN_STATS, metadataColumnStatsRDD);
     }
+    if (enabledPartitionTypes.contains(MetadataPartitionType.PARTITION_STATS)) {
+      final HoodieData<HoodieRecord> partitionStatsRDD = convertMetadataToPartitionStatsRecords(commitMetadata, context, dataMetaClient, metadataConfig);
+      partitionToRecordsMap.put(MetadataPartitionType.PARTITION_STATS, partitionStatsRDD);
+    }
```
A side topic, does this compose a sub-graph in DAG without triggering the execution?  Ideally, all types of metadata should be computed in parallel leveraging the parallelism in Spark, instead of being computed type by type sequentially.

codope replied:
No. That's a very good point! Typically, we are enabling indexes one at a time. But, I agree with you. I will work on it in HUDI-7690. Btw, this limitation is only for initialization. In case of update, we don't compute sequuentially.

yihua commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1872,4 +1883,175 @@ public HoodieRecord next() {
   }
+
+  public static HoodieData<HoodieRecord> convertFilesToPartitionStatsRecords(HoodieEngineContext engineContext,
+                                                                             List<DirectoryInfo> partitionInfoList,
```
Does this contain log files in MOR tables?

codope replied:
No, the assumptions of colstats apply for parition stats too. Log files are considered in subsequent updates with the write status. So, just like colstats, we will have to compact the table before enabling partition stats.

yihua replied:
I see. Is this for initializing the partition stats only?  Do we have a guard to throw an exception if the colstats or partition stats partition is initialized on the latest file slices with log files?

codope replied:
Yes, this is just for initialization. We just log a warning and continue for other filegroups. This is what colstats is also doing. However, given that partition stats will be enabled by default, I think it makes sense to throw exception. Will do it.

yihua commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1872,4 +1883,175 @@ public HoodieRecord next() {
+    return readColumnRangeMetadataFrom(filePartitionPath, datasetMetaClient, columnsToIndex);
+  }
+
+  public static HoodieData<HoodieRecord> convertMetadataToPartitionStatsRecords(HoodieCommitMetadata commitMetadata,
```
Can this method be consolidated with `convertFilesToPartitionStatsRecords`? They share common code.

codope replied:
High-level steps are same but the two methods work with different objects. This method works with writeStats while `convertFilesToPartitionStatsRecords` works with actual file itself. Also, this method has some preprocessing which is not required in the other method. I think the code reads better inline.

yihua replied:
Got it, makes sense.  We can revisit later why it needs two types of methods/code path for generating partition stats records (track it in JIRA).

codope replied:
Ack, HUDI-7696 to track.

yihua commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -67,6 +70,61 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+        .reduce(BaseFileUtils::mergeRanges).get();
+  }
+
+  private static <T extends Comparable<T>> HoodieColumnRangeMetadata<T> mergeRanges(HoodieColumnRangeMetadata<T> one,
```
Do we merge column ranges between the base and log files (e.g., log file contains updates and deletes)?  Is that covered in this PR or by existing column stats logic (how does that work)?

yihua replied:
Got it.  I'm trying to understand how col stats are merged between the base and log files, e.g., for custom payload, because we may not be able to simply take the minimum of all minimum values or the maximum of all maximum values from the base and log files, as there can be deletes and there can be event-time based merging.  We can discuss this in a separate thread.

codope replied:
We follow the same approach as for colstats - https://github.com/apache/hudi/blob/f99b181a04e2ee549009d38deca7597219c4ce46/hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java#L372
If the new record is for delete, then we ignore the previous one. If previous record is for delete, then we take the newer one, otherwise just merge the stats. I don't think we do event-time based merging for MDT.

yihua commented on hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/functional/TestDataSkippingWithMORColstats.java:
```diff
@@ -318,6 +321,7 @@ private Map<String, String> getOptions() {
     options.put(HoodieMetadataConfig.ENABLE.key(), "true");
     options.put(HoodieMetadataConfig.ENABLE_METADATA_INDEX_COLUMN_STATS.key(), "true");
     options.put(HoodieMetadataConfig.COLUMN_STATS_INDEX_FOR_COLUMNS.key(), "trip_type");
+    // options.put(HoodieMetadataConfig.ENABLE_METADATA_INDEX_PARTITION_STATS.key(), "false");
```
nit: remove?

yihua commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndex.scala:
```diff
@@ -0,0 +1,191 @@
+ * Test cases on partition stats index with Spark datasource.
+ */
+@Tag("functional")
+class TestPartitionStatsIndex extends PartitionStatsIndexTestBase {
```
Could you add docs on each test scenario so it's easy to read?

codope commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -67,6 +70,61 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+   * @param fileColumnRanges List of column range statistics for each file in a partition
+   */
+  public static <T extends Comparable<T>> HoodieColumnRangeMetadata<T> getColumnRangeInPartition(@Nonnull List<HoodieColumnRangeMetadata<T>> fileColumnRanges) {
+    if (fileColumnRanges.size() == 1) {
```
`fileColumnRanges` should not be empty. For empty parquet files, we will err out before this point. I'll add a validation anyway.

codope commented on hudi-common/src/main/java/org/apache/hudi/common/util/BaseFileUtils.java:
```diff
@@ -67,6 +70,61 @@ public static BaseFileUtils getInstance(HoodieFileFormat fileFormat) {
+    }
+    // There are multiple files. Compute min(file_mins) and max(file_maxs)
+    return fileColumnRanges.stream()
+        .sequential()
```
does not matter.. removed it.

codope commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -403,6 +400,13 @@ private boolean initializeFromFilesystem(String initializationTime, List<Metadat
             ValidationUtils.checkState(functionalIndexPartitionsToInit.size() == 1, "Only one functional index at a time is supported for now");
             fileGroupCountAndRecordsPair = initializeFunctionalIndexPartition(functionalIndexPartitionsToInit.iterator().next());
             break;
+          case PARTITION_STATS:
+            if (dataWriteConfig.getColumnsEnabledForColumnStatsIndex().isEmpty()) {
+              LOG.warn("Skipping partition stats index initialization as target columns are not set");
+              continue;
+            }
+            fileGroupCountAndRecordsPair = initializePartitionStatsIndex(partitionInfoList);
+            break;
```
Agree. Will do it in HUDI-7691

yihua commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndexWithSql.scala:
```diff
@@ -0,0 +1,106 @@
+
+      // Test insert into
+      spark.sql(s"insert into $tableName values(2, 'a2', 10, 1000, cast('$partitionValue' as date))")
+      checkAnswer(s"select _hoodie_record_key, _hoodie_partition_path, id, name, price, ts, cast(dt as string) from $tableName order by id")(
```
I don't seem to see the validation on the partition pruning, e.g., in the logical plan, in the tests in this class.

codope replied:
I have it in two seprate tests - `testPartitionStatsWithPartitionBy` and `testPartitionStatsIndexFilePruning`

yihua reviewed (approved):
+1 you can land the PR once all clarifying questions are answered.

KnightChess reviewed (commented):
hello, I found partition stats will not effort, the current implementation does not seem to achieve the effect of partition filtering.
- first

in this picture, I change the ut filter to trigger partition stats index.
![image](https://github.com/apache/hudi/assets/20125927/5a62577c-af27-49d7-9e24-a350dc3b0b8b)

partition_stats will not save fileName, so if reuse `CSI` logical, it will throw null point in group by key
![image](https://github.com/apache/hudi/assets/20125927/19f53b39-ef29-4273-8e85-2bddc4998023)
![image](https://github.com/apache/hudi/assets/20125927/c40858d0-22be-44f0-810a-cfd8a5f51187)

- second

and have a question, I am not sure this pr is use to `partition` purge like physical partition col, which mean use other field min/max to get which physical partitions to list fileSlice. or filter fileName like `CSI`, `RLI`.
thanks

KnightChess commented:
creaete a issue to track it https://issues.apache.org/jira/browse/HUDI-7829
